### PR TITLE
Support log_model with code model in langchain

### PR DESCRIPTION
**Pull request by annzhang-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/annzhang-db/mlflow/pull/11817?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11817/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11817
```

</p>
</details>

### Related Issues/PRs

Work to follow:
- Introduce ModelConfig class https://github.com/mlflow/mlflow/pull/11841
- Using model_config instead of code_paths[0] for the config file path, move model_config to be under model_code https://github.com/mlflow/mlflow/pull/11843
- Code validation on the provided file path: warn on magic commands, port over dbutils logic https://github.com/mlflow/mlflow/pull/11844
- mlflow.models.set_model(): validate object type https://github.com/mlflow/mlflow/pull/11842 

### What changes are proposed in this pull request?
- Remove restriction for file to be called "chain.py"
- Add logic to read from file or databricks notebook
- Introduce `model_code` as separate from `code`

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/langchain/test_langchain_model_export.py [hunk 1]
```diff
@@ -2615,7 +2601,6 @@ def test_save_load_chain_as_code_optional_code_path():
             artifact_path="model_path",
             signature=signature,
             input_example=input_example,
-            code_paths=[],
```
Comment (annzhang-db) [inline tests/langchain/test_langchain_model_export.py:2618]: test code_paths=None

File: mlflow/langchain/__init__.py [hunk 2]
```diff
@@ -246,11 +251,11 @@ def load_retriever(persist_directory):
             formatted_code_path.append(lc_model)
         else:
             raise mlflow.MlflowException.invalid_parameter_value(
-                f"If the {lc_model} is a string, it must be a valid python "
+                f"If the provided model '{lc_model}' is a string, it must be a valid python "
                 "file path containing the code for defining the chain instance."
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:255]: For the case, can we log it as a separate model metadata like `model_code_path` instead of reusing "code_paths" ?
Because it has differences with existing `code_paths` files (its code is not a common module, but it should contain `set_chain` method, and it might come from databricks notebook),
and recently we are expanding MLflow `code_paths` functionality, and the newly added `model_code_path` can't support the expanded `code_paths` functionality and make code messy.

I have some related discussion in this doc:
https://docs.google.com/document/d/144wAwgXsQ40C3dDsoObX0LfXp33aRbvEVCCq-VYgJqw/edit#bookmark=id.d1xwn4gq68ce

https://docs.google.com/document/d/144wAwgXsQ40C3dDsoObX0LfXp33aRbvEVCCq-VYgJqw/edit#bookmark=id.w62724445pwk

CC @BenWilson2

File: mlflow/langchain/utils.py [hunk 3]
```diff
@@ -278,13 +309,33 @@ def _validate_and_wrap_lc_model(lc_model, loader_fn):
...
+                return lc_model
+        except Exception:
+            try:
+                from databricks.sdk import WorkspaceClient
+                from databricks.sdk.service.workspace import ExportFormat
+
+                w = WorkspaceClient()
+                response = w.workspace.export(path=lc_model, format=ExportFormat.SOURCE)
+                decoded_content = base64.b64decode(response.content)
+                # TODO: code validation
```
Comment (WeichenXu123) [inline mlflow/langchain/utils.py:331]: I think validation can be done in this way: starting a child process, run this python file, and then check if `__databricks_rag_chain__` is set to a valid langchain model. we can do it in follow-up PR.

File: mlflow/langchain/__init__.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/langchain/__init__.py:255]: Yes, we should keep this logic separate. This main entry point to a chain definition should be handled distinctly to avoid having to add error handling logic that would need to be applied to branching decision logic if this were overloaded into code_paths for directory traversal for dependent relative and absolute import statements with dependency inference. 

Regardless of this fact, what is the mechanism for handling dependent imports within this implementation? If a user has external imports to custom code that rely on absolute imports, will this notebook path preserve its directory structure from the workspace root?

Comment (annzhang-db) [reply mlflow/langchain/__init__.py]: Updated the PR to separate this logic. There isn't any mechanism for handling dependent imports right now, and the notebook path does not preserve its directory structure. Should this be a requirement here?

File: mlflow/langchain/utils.py (hunk 3 above)
Comment (annzhang-db) [reply mlflow/langchain/utils.py:332]: Yes, the code validation won't be in this PR.

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -36,7 +36,12 @@
     _update_langchain_model_config,
     patched_inference,
 )
-from mlflow.langchain._rag_utils import _CODE_CONFIG, _CODE_PATH, _set_config_path
+from mlflow.langchain._rag_utils import (
+    _CODE_CONFIG,
+    _CODE_PATH,
+    _set_chain,
+    _set_config_path,
+)
```
Comment (harupy) [inline mlflow/langchain/__init__.py:44]: Is it possible to revert unnecessary changes?

File: mlflow/langchain/utils.py [hunk 5]
```diff
@@ -278,13 +309,35 @@ def _validate_and_wrap_lc_model(lc_model, loader_fn):
...
-                f"If {lc_model} is a string, it must be the path to a file "
-                "named `chain.py` on the local filesystem."
+                f"If the provided model '{lc_model}' is a string, it must be a valid python "
+                "file path or a databricks notebook file path containing the code for defining "
+                "the chain instance."
             )
-        return lc_model
+
+        try:
+            with open(lc_model) as _:
```
Comment (harupy) [inline mlflow/langchain/utils.py:322]: What is this `open` for?

File: tests/langchain/test_langchain_model_export.py [hunk 6]
```diff
@@ -2374,7 +2380,53 @@ def test_save_load_chain_as_code():
...
+    # Read the contents of the original chain file
+    with open("tests/langchain/chain.py") as chain_file:
+        chain_file_content = chain_file.read()
+
+    with tempfile.NamedTemporaryFile(mode="w+", delete=True, suffix=".py") as temp_file:
+        # Add "import dbutils" to temporary chain file
+        temp_file.write(chain_file_content)
+        temp_file.flush()
+
+        chain_path = temp_file.name
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:2402]: we can use the `tmp_path` fixture and store the file there.

File: mlflow/langchain/__init__.py [hunk 7]
```diff
@@ -236,21 +243,22 @@ def load_retriever(persist_directory):
...
+                f"If the provided model '{lc_model}' is a string, it must be a valid python "
+                "file path or a databricks notebook file path containing the code for defining "
+                "the chain instance."
             )
 
-        if len(code_paths) > 1:
+        if code_paths and len(code_paths) > 1:
             raise mlflow.MlflowException.invalid_parameter_value(
                 "When the model is a string, and if the code_paths are specified, "
                 "it should contain only one path."
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:264]: The error message should be updated because we have separate "model code" logic

File: mlflow/langchain/utils.py (hunk 5 above)
Comment (annzhang-db) [reply mlflow/langchain/utils.py:322]: This is to determine if the lc_model is a normal file path or databricks notebook. I couldn't find a better way to do this, open to any suggestions.

File: mlflow/langchain/__init__.py [hunk 8]
```diff
@@ -536,6 +550,19 @@ def load_retriever(persist_directory):
     )
 
 
+@experimental
+def set_chain(chain):
```
[the remark this answers is not in the document]
Comment (annzhang-db) [reply mlflow/langchain/__init__.py]: Discussed with Ben, we will not be using set_chain. Instead going to implement mlflow.models.set_model.

File: mlflow/langchain/__init__.py (hunk 7 above)
Comment (annzhang-db) [reply mlflow/langchain/__init__.py]: Sorry, not sure what exactly needs to be updated here, but this exception raised from code_paths will be entirely removed in the next PR so there shouldn't be any problems here.

File: mlflow/langchain/__init__.py [hunk 9]
```diff
@@ -814,9 +822,10 @@ def _load_model_from_local_fs(local_model_path):
             config_path = None
 
         flavor_code_path = flavor_conf.get(_CODE_PATH, "chain.py")
+        flavor_model_code_config = flavor_conf.get(FLAVOR_CONFIG_MODEL_CODE)
         code_path = os.path.join(
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:826]: nit:
```suggestion
        model_code_path = os.path.join(
```

File: mlflow/langchain/__init__.py [hunk 10]
```diff
@@ -814,9 +822,10 @@ def _load_model_from_local_fs(local_model_path):
             config_path = None
 
         flavor_code_path = flavor_conf.get(_CODE_PATH, "chain.py")
+        flavor_model_code_config = flavor_conf.get(FLAVOR_CONFIG_MODEL_CODE)
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:825]: Forward comment: https://github.com/mlflow/mlflow/pull/11843#discussion_r1582536788

File: mlflow/langchain/__init__.py [hunk 11]
```diff
@@ -256,7 +258,8 @@ def load_retriever(persist_directory):
                 f"Current code paths: {code_paths}"
             )
 
-    code_dir_subpath = _validate_and_copy_code_paths(formatted_code_path, path)
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+    model_code_dir_subpath = _validate_and_copy_model_code_path(model_code_path, path)
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:262]: Forward comment: https://github.com/mlflow/mlflow/pull/11843#discussion_r1582520252

File: mlflow/langchain/__init__.py [hunk 12]
```diff
@@ -233,21 +235,21 @@ def load_retriever(persist_directory):
...
             raise mlflow.MlflowException.invalid_parameter_value(
-                f"If the {lc_model} is a string, it must be a valid python "
-                "file path containing the code for defining the chain instance."
+                f"If the provided model '{lc_model}' is a string, it must be a valid python "
+                "file path or a databricks notebook file path containing the code for defining "
+                "the chain instance."
             )
-
-        if len(code_paths) > 1:
+        if code_paths and len(code_paths) > 1:
```
Comment (annzhang-db) [inline mlflow/langchain/__init__.py:252]: https://github.com/mlflow/mlflow/pull/11844#discussion_r1583423605

Comment (annzhang-db) [reply mlflow/langchain/__init__.py:252]: This will get removed in the next PR: https://github.com/mlflow/mlflow/pull/11843/files

File: mlflow/langchain/__init__.py [hunk 13]
```diff
@@ -256,7 +258,8 @@ def load_retriever(persist_directory):
                 f"Current code paths: {code_paths}"
             )
 
-    code_dir_subpath = _validate_and_copy_code_paths(formatted_code_path, path)
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+    model_code_dir_subpath = _validate_and_copy_model_code_path(model_code_path, path)
 
     if signature is None:
```
Comment (annzhang-db) [inline mlflow/langchain/__init__.py:264]: https://github.com/mlflow/mlflow/pull/11844#discussion_r1583428715

File: mlflow/langchain/__init__.py [hunk 14]
```diff
@@ -330,12 +335,14 @@ def load_retriever(persist_directory):
...
         predict_stream_fn="predict_stream",
         streamable=streamable,
+        model_code=model_code_dir_subpath,
         **model_data_kwargs,
     )
 
     if Version(langchain.__version__) >= Version("0.0.311"):
         checker_model = lc_model
         if isinstance(lc_model, str):
+            # TODO: use model_config instead of code_paths[0]
```
Comment (annzhang-db) [inline mlflow/langchain/__init__.py:345]: https://github.com/mlflow/mlflow/pull/11844#discussion_r1583433977

File: mlflow/langchain/utils.py [hunk 15]
```diff
@@ -278,13 +309,35 @@ def _validate_and_wrap_lc_model(lc_model, loader_fn):
...
+                return lc_model
+        except Exception:
+            try:
+                from databricks.sdk import WorkspaceClient
+                from databricks.sdk.service.workspace import ExportFormat
+
+                w = WorkspaceClient()
```
Comment (annzhang-db) [inline mlflow/langchain/utils.py:329]: https://github.com/mlflow/mlflow/pull/11844#discussion_r1583435836

File: mlflow/utils/model_utils.py [hunk 16]
```diff
@@ -162,6 +163,20 @@ def _validate_and_copy_code_paths(code_paths, path, default_subpath="code"):
     return code_dir_subpath
 
 
+def _validate_and_copy_model_code_path(code_path, path, default_subpath="model_code"):
```
Comment (annzhang-db) [inline mlflow/utils/model_utils.py:166]: https://github.com/mlflow/mlflow/pull/11844#discussion_r1583452851

File: mlflow/langchain/__init__.py (hunk 10 above)
Comment (annzhang-db) [reply mlflow/langchain/__init__.py:825]: Yes, removed in this PR https://github.com/mlflow/mlflow/pull/11843/files but need to port it over 😓

File: mlflow/langchain/__init__.py (hunk 13 above)
Comment (annzhang-db) [reply mlflow/langchain/__init__.py:264]: Resolve signature issues with https://github.com/mlflow/mlflow/pull/11817/files#r1583519300 this potentially

File: mlflow/utils/model_utils.py (hunk 16 above)
Comment (annzhang-db) [reply mlflow/utils/model_utils.py:166]: This is reworked in the next PR https://github.com/mlflow/mlflow/pull/11843, let's address it there

**Review by sunishsheth2009 — APPROVED** [review]

**Review by sunishsheth2009 — COMMENTED** [review]
Approving this so we can merge this and we can address the TODOs so we can do it in small batch
